### ISSUE TITLE
Production Sub Dependency Excluded If Also Included in devDependencies (Node 8.10)

### ISSUE THREAD

**Issue by AmberDean** [serverless/serverless#4963]
I upgraded to Node 8.10, and after attempting a deployment, found out that graphql was missing from my node_modules folder.  I'm guessing that it got stripped out because it was in my devDependencies list, but it was also a production sub-dependency for another module.

I did not have this issue with Node 6.10.
I verified this is happening in both my local windows environment and my linux CI environment.
Moving the missing module from devDependencies to dependencies in package.json resolved the issue and is my current workaround.

Similar or dependent issues:
* #12345

## Additional Data

* ***Serverless Framework Version you're using***: 1.27.1
* ***Operating System***: Linux + Windows\

**Comment by AmberDean** [comment]
Also worth mentioning, the only plugin being used is serverless-plugin-typescript.

**Comment by AmberDean** [comment]
@HyperBrain I can verify that it was only the change of the Node version that caused the problem.  And yes, I'm actually in the process right now of trying out serverless-webpack.  I found that serverless-plugin-type seemed to work better for me for offline development.

**Comment by HyperBrain** [comment]
The missing dependency with your current setup might be related to a change of the NPM version and Serverless removing the wrong folder from the node_modules tree. Which version of NPM did you use before and after the update to Node 8.10? NPM >= 3.x does module flattening within the node_modules folder, so a single dependency is not stored within it's parent anymore, but on the root level.
I think there is a bug in the Serverless packaging, that removes this flattened module if it is also present in devDependencies, although it does not get installed at root level by the main project, but as a second level dependency. Serverless thinks by mistake that the module is a dev dependency because it seems not to handle deep level deps correctly.

**Comment by AmberDean** [comment]
I was using NPM 5.6.0 both before and after the update to Node 8.10

**Comment by HyperBrain** [comment]
Ok. Then the NPM version should not have any effect as it was the same.

**Comment by Maxwell2022** [comment]
I actually have a very similar issue. Using `sequelize` and `pg`. `pg` is in my `dependencies` but once packaged it's not in the node module. I'm using `serverless-webpack` with `ts-loader` and `webpack v4`.

my config look like this:
```
custom:
  defaultConfigType: 'cloud'
  webpack:
    packager: yarn
    webpackConfig: './webpack/config.${env:WEBPACK_CONFIG_TYPE, self:custom.defaultConfigType}.js'
    includeModules: true
  serverless-offline:
    port: 4000
    host: '0.0.0.0'

package:
  individually: true
``` 

my webpack config:

```
const path = require('path');
const nodeExternals = require('webpack-node-externals');
const slsw = require('serverless-webpack');

module.exports = {  
  entry: slsw.lib.entries,
  mode: slsw.lib.webpack.isLocal ? 'development': 'production',
  devtool: 'source-map',
  target: 'node',
  externals: [nodeExternals()],
  optimization: {
    // We no not want to minimize our code.
    minimize: false
  },
  performance: {
    // Turn off size warnings for entry points
    hints: false
  },
  module: {
    rules: [
      { test: /\.ts(x?)$/, use: [{ loader: 'ts-loader' }], exclude: /node_modules/ },
      { test: /\.graphql?$/, use: [{ loader: 'webpack-graphql-loader' }]},
    ],
  },
  resolve: {
    extensions: ['.ts', '.js', '.tsx', '.jsx']
  },
  output: {
    libraryTarget: 'commonjs2',
    path: path.join(__dirname, './../.build'),
    filename: '[name].js',
    sourceMapFilename: '[file].map',
    devtoolModuleFilenameTemplate: '../../[resource-path]'
  }
};
```

I'm also using the offline plugin and everything is working well offline (because there is no packaging). I'm not really sure what would be the process to debug this. I've noticed that for each Function, it's generating a folder and inside this folder there is a `package.json` and `yarn.lock`.

The dependencies of this `package.json` is different than mine. The `pg` module is also missing within the `dependencies` folder generated by webpack I think.

I keep investigating this issue...

**Comment by HyperBrain** [comment]
Hi @Maxwell2022 , especially using `sequelize` you have to force include the database modules (`pg` in your case).

Please add/change the following in your sls.yml:
```
custom:
  webpack:
    ...
    includeModules:
      forceInclude:
        - pg
```

**Comment by Maxwell2022** [comment]
Yes, Sorry I didn't update this issue. I spent something like 2 hours debugging it... That's what I ended up doing tho.

It's really miss-leading because there is no error message, or even warning saying that the module that I **explicitly have in my `dependencies`** is not going to be installed by webpack. I think it would be very beneficial to display the list of modules that will not be installed (even tho they are in the dependencies) because they are not used or dynamically imported. That would either indicate to the developer that he needs to clean their dependencies or force install them.

I should Probably open an issue in this specific project tho

Thanks for the reply @HyperBrain

**Comment by Maxwell2022** [comment]
I think this issue can be closed :)

**Comment by pmuens** [comment]
Closing this issue since it's resolved.

**Comment by kasvtv** [comment]
This shouldn't have been closed as this is still an issue with `1.67.0`. For me it happens because my dev dependency `jest` has a transitive dependency `graceful-fs` and I have another prod dependency that also uses `graceful-fs`. It doesn't end up in the package and my code fails due to the `graceful-fs` module not being found.
